Test Playing Cards v2

miketwo opened the pull request:
New PR for the local branch. Now you should be able to `git fetch --all` and `git checkout test_playing_cards`

[Old PR for reference](https://github.com/Vesper-arch/Slay-The-Spire-in-Python/pull/34).

Vesper-arch commented:
After about 35 bug fixes, this showed up and I don't know what is causing it or how to fix it. I know it has something to do with the input function. This output continues to repeat indefinitely.
![Screenshot 2024-01-10 18 29 08](https://github.com/Vesper-arch/Slay-The-Spire-in-Python/assets/137581120/f80c0826-67dc-43a7-bb71-6896729dc3c6)

miketwo commented:
That's a lot of fixes, wow....

Vesper-arch commented:
The way you handled the inputs by choosing the first one worked for a while but it went into a loop when the first options didn't meet the conditions.

Vesper-arch commented:
I actually used to do that but I found that the inputted option(which is subtracted by 1 to make it 0-based) didn't reflect the place in the real list so I greyed out the invalid options.

Vesper-arch commented:
Man, those 44 lines of code uncovered every little mistake I made...

miketwo commented:
lol - automated testing is powerful. I'll take a look at the infinitely repeating output now...

miketwo commented:
Got it. Should be failing on the next thing now.

miketwo commented:
Fixed the infinite loop for real now. (Thought I had fixed it but not quite. The randomness makes it a little challenging to reproduce errors.)

The one I'm hitting the most now is that the boss dies midway through a massive `use_fiendfire` attack, but the attack continues. The boss "dies" a second time, and the game tries to remove it from `active_enemies`, but it's already gone, so it crashes. This would probably happen during any mutli-step attack where the boss dies in the middle.

Lines from the test output:
```
You dealt 5 damage(0 Blocked) to Slime Boss
Affected by: 
Weak(x0.75 dmg) | Vulnerable(x1.75 dmg) | 3 Strength | The Boot(dmg increased to 5)
Slime Boss has died.              <------------ once
Feed was Exhausted.
You dealt 5 damage(0 Blocked) to Slime Boss
Affected by: 
Weak(x0.75 dmg) | Vulnerable(x1.75 dmg) | 3 Strength | The Boot(dmg increased to 5)
Slime Boss has died.              <------------ twice
...
>       active_enemies.remove(self)
E       ValueError: list.remove(x): x not in list
```

This could be a bigger question of when to check if something is dead, and what to do during multi-attacks in that case.

Once we fix this it might be good to artificially inflate the boss's health so the tests can continue.

miketwo commented:
Fixed a bunch more. The two I'm stuck on are similar:

- Boss dies twice during multi-attach, so remove from active_enemies fails
- A card ("Feed") is Exhausted twice from Double Tap, and the second move to the exhaust pile fails.

That's it for tonight...

Vesper-arch commented:
My proposed solution to the infinite input loop is to write a new version of `list_input` that has the conditions built-in, this will shorten the code, make the invalid input messages more accurate, and the test can loop over the `array` parameter and see the first one that meets the conditions.

miketwo commented:
Probably a better long-term solution but infinite loop should be fixed now. Let me push what I've got. It was a loooooooot of fixes. But I think it passes the ubertest now.

Vesper-arch commented:
It works!!

Vesper-arch commented on items.py:
```diff
@@ -254,14 +254,19 @@ def use_dropkick(targeted_enemy, using_card, entity):
+        view.view_piles(pile=view_cards, entity=entity)
+        option = view.list_input(f"Choose a card to make {'a copy' if not using_card.get('Upgraded') else '2 copies'} of > ", valid_cards)
+        #convert option to index of entity.hand
+        option = valid_cards[option][0]
```
I don't get how the conversion works here. Can you explain this?

miketwo replied:
```python
# enumerate returns an index and item for items in a list, so you don't have to manage a counter yourself.
# So this line will create a new list that has the indicies of the hand and the card if it's valid.
# Ex: If there's 3 cards in the hand, at indicies 0, 1, 2 -- and the middle one is invalid -- this would create the following list:
# [(0, first_card), (2, third_card)]
# Note that the indexes of of this new list are different from the original. 0 and 1 map to 0 and 2.
valid_cards = [(idx, card) for idx, card in enumerate(entity.hand) if card.get('Type') in ('Attack', 'Power')]

# We pull out the cards part of the tuple and put that up for display
view_cards = [card[1] for card in valid_cards]
view.view_piles(pile=view_cards, entity=entity)

# Then, once we select a card, we know how to map the indicies. The chosen index maps to `valid_cards `, and from there, the first part of the tuple matches the index in the hand.
option = view.list_input(..., valid_cards)
